### ISSUE TITLE
Gtk SourceView Support

### ISSUE THREAD

**Issue by Blquinn** [GtkSharp/GtkSharp#168]
Hello, thanks for all the hard work you've put into this binding.

Would it be possible to add a binding to GtkSourceView to this lib? Would this lib be a good place to put that?

Could someone give me a high level overview of how to create a binding for SV myself?

Thank you!

**Comment by Dadoum** [comment]
You can use gapi3 to bind gtk libraries, and mono made a tutorial on how to use GAPI. It will just require a glue library. 

I already binded GtkSourceView-4 in GtkSharp so it works.

**Comment by Blquinn** [comment]
@Dadoum Thanks for the info! I will look into doing that. Do you think it's worth while to maintain a binding somewhere?

**Comment by Dadoum** [comment]
Yes I think it should be somewhere for everyone but surely not in this repository. 
Anyway if you need it it is available [there](https://github.com/Dadoum/iCode/tree/develop/Source/iCode/Native/GtkSource) (glue is in root of the project, and Std.cs file is required) but I am not going to create repository for it because I lost generation scripts...

**Comment by Blquinn** [comment]
So looking a bit at your project an GtkSharp. It seems like I need to
1. Build the GapiCodegen and GapiFixup tools
2. Write the xml .metadata files by hand
3. Write a cake script to run Fixup on .metadata to generate the .api files, then run Codegen on the api files

Is that about right? Also, where is the source for the glue library, or how do I generate that?

**Comment by harry-cpp** [comment]
1. Build gapi parser
2. Run gapi parser for the lib you want to generate stuff (this should help: https://www.mono-project.com/docs/gui/gtksharp/gapi/)
3. Add the xml file
4. Write extra metadata files / code for fixing up stuff
5. Modify CAKE script for the extra lib to the [AssemblyList](https://github.com/GtkSharp/GtkSharp/blob/develop/CakeScripts/Settings.cake#L12)
6. Profit?

You could see a lot of what you'd need to do in this commit: https://github.com/GtkSharp/GtkSharp/commit/768662ec7775ea795a5678b08e3881dbb0324e06

**Comment by Blquinn** [comment]
@harry-cpp Sounds like your method would add GtkSource to this repo? Dadoum suggested to put it somewhere else. Seems like putting it in this repo would be the easiest, considering most of the deps live here. Happy to try either way, just let me know what you think. Most profits for all if I can contribute something back to this lib I think :)

**Comment by Dadoum** [comment]
I think this repo should stay as the minimal setup for GtkSharp. I think it would be better to create another repo for other libraries.

**Comment by Blquinn** [comment]
@Dadoum I would accept that answer, but also mention that it could be a separate nuget package, but still live in this repo. So it wouldn't necessarily bloat the base package, if that's what you're worried about.

**Comment by Dadoum** [comment]
Could be a solution

**Comment by hol430** [comment]
Hi all, just chiming in here to say that I generated some sourceview bindings a few weeks ago, and pushed them up [here](https://github.com/hol430/SourceView). It's all a bit of a mess at the moment, and I'm not really much of an expert on this stuff but I'm hoping to get back to it at some point and it seems to be working for me (on Linux anyway).

Do any of you guys have any bindings at the moment? If so I'm happy to delete mine and switchto using yours, to avoid any duplication of effort. A nuget package would be really convenient, so I'm happy to help out where I can to achieve that.

**Comment by Blquinn** [comment]
@hol430 Dadoum has a binding, but he lost his generation scripts, so I think yours may be the best we have at this point.

I'm still unclear on whether we should be forking GtkSharp and adding this to Libs, or making this its own repo.

I'll give your binding a shot and see if it works for me.

What specifically do you think is a mess about yours?

**Comment by hol430** [comment]
Nothing too major, I just haven't tested it all that thoroughly, want to change some of the namespace stuff to make it more consistent with this repo, and I forgot to record which version of the gtksourceview code I used to generate the bindings.

@harry-cpp (I assume you're the maintainer for this repo) - would you prefer to see sourceview bindings in this repo, or somewhere else?

**Comment by harry-cpp** [comment]
I would totally be up for them being in this repo.

**Comment by Blquinn** [comment]
@hol430 Sooooo......... You're on it? :joy:

Seriously though, I'm happy to give what little assistance I can offer. Seems like testing is a priority for you.

**Comment by hol430** [comment]
Yep, I'll see if I can integrate it into this repo and create a pull request. 768662e is a great starting point, thanks.

**Comment by Dadoum** [comment]
Also which gapi-parser are you all using ? I am using the one in the gtk-sharp-3 package, but it seems to not provide everything written in the headers (the problem is visible in Scintilla library, it misses some structs), it has problems with CPP mixed within the header.

**Comment by hol430** [comment]
I tried using the gapi3-* tools from the gtk-sharp-3 package, but I was having problems which sound similar to yours. gapi3-codegen was giving errors along the lines of "missing GetType method" in some interfaces. Compiling and using this repo's version of gapi3-parser corrected the problem for me, although the gapi3-codegen tool seems to be doing some strange things which I'm still looking into.

On a side note, what is the intended/correct way of compiling the gapi3-parser in this repo? I tried using meson but couldn't get it to work, and in the end just gave up and used mcs to compile, and tweaked the code so it could find the .pl files. If someone could advise me on how to build this via meson (or other tools) that'd be great.
